File: problems/930_binary_subarrays_with_sum/solution.py
from typing import Dict, List


class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        count: Dict[int, int] = {0: 1}
        prefix = 0
        result = 0
        for num in nums:
            prefix += num
            result += count.get(prefix - goal, 0)
            count[prefix] = count.get(prefix, 0) + 1
        return result

File: problems/930_binary_subarrays_with_sum/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int, int> count;
        count[0] = 1;
        int prefix = 0, result = 0;
        for (int num : nums) {
            prefix += num;
            result += count[prefix - goal];
            count[prefix]++;
        }
        return result;
    }
};
